services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,377 @@
+
+		return toReturn, nil
+	}
+}
+
+// buildObsDepthWithIntrinsics will use the methodology in Manduchi et al. to find obstacle points
+// before clustering and projecting those points into 3D obstacles.
+func (o *obsDepth) buildObsDepthWithIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		// Check if we have intrinsics here or in the camera properties. If not, don't even try
+		props, err := src.Properties(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if props.IntrinsicParams != nil {
+			o.intrinsics = props.IntrinsicParams
+		}
+		if o.intrinsics == nil {
+			return nil, errors.New("tried to build obstacles depth with intrinsics but no instrinsics found")
+		}
+
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		o.dm = dm
+		o.makePointList(sampleN)
+
+		// Use some goroutines to determine if each depth pixel is an obstacle
+		doneCh := make(chan bool, len(o.ptChunks))
+		var wg sync.WaitGroup
+		for i, chunk := range o.ptChunks {
+			wg.Add(1)
+			go func(i int, chunk []obsPoint) {
+				defer wg.Done()
+				for j, p := range chunk {
+					o.ptChunks[i][j].isObstacle = o.isObstaclePoint(p.point)
+				}
+				doneCh <- true
+			}(i, chunk)
+		}
+		go func() {
+			wg.Wait()
+			close(doneCh)
+		}()
+		wg.Wait()
+
+		// Cluster on the 2D depth points and then project the 2D clusters into 3D boxes
+		outClusters, err := o.performKMeans(o.k)
+		if err != nil {
+			return nil, err
+		}
+		boxes, err := o.clustersToBoxes(outClusters)
+		if err != nil {
+			return nil, err
+		}
+
+		// Packaging the return depending on if they want PCDs
+		n := int(math.Min(float64(len(outClusters)), float64(len(boxes)))) // should be same len but for safety
+		toReturn := make([]*vision.Object, n)
+		for i := 0; i < n; i++ { // for each cluster/box make an object
+			if o.returnPCDs {
+				pcdToReturn := pointcloud.New()
+				basicData := pointcloud.NewBasicData()
+				for _, pt := range outClusters[i].Observations {
+					if len(pt.Coordinates()) >= 3 {
+						vec := r3.Vector{X: pt.Coordinates()[0], Y: pt.Coordinates()[1], Z: pt.Coordinates()[2]}
+						err = pcdToReturn.Set(vec, basicData)
+						if err != nil {
+							return nil, err
+						}
+					}
+				}
+				toReturn[i] = &vision.Object{PointCloud: pcdToReturn, Geometry: boxes[i]}
+			} else {
+				toReturn[i] = &vision.Object{Geometry: boxes[i]}
+			}
+		}
+		return toReturn, nil
+	}
+}
+
+// makePointList will grab points from the depthmap. Sample every n.
+func (o *obsDepth) makePointList(n int) {
```

Comment:
I think this method may be unnecessary. You can consolidate many operations in the first pass through the depth map.